packages/indy-vdr/src/IndyVdrApi.ts

```diff
@@ -0,0 +1,71 @@
+import type { Key } from '@aries-framework/core'
+import type { EndorseDidTxAction } from '@aries-framework/core/src/modules/dids/types'
+import type { AttribRequest, IndyVdrRequest } from '@hyperledger/indy-vdr-shared'
+
+import { AgentContext, injectable, TypedArrayEncoder } from '@aries-framework/core'
+
+import { verificationKeyForIndyDid, parseIndyDid } from './dids/didIndyUtil'
+
+@injectable()
+export class IndyVdrApi {
+  private agentContext: AgentContext
+
+  public constructor(agentContext: AgentContext) {
+    this.agentContext = agentContext
+  }
+
+  private async multiSignRequest<Request extends IndyVdrRequest>(
+    request: Request,
+    signingKey: Key,
+    identifier: string
+  ) {
+    const signature = await this.agentContext.wallet.sign({
+      data: TypedArrayEncoder.fromString(request.signatureInput),
+      key: signingKey,
+    })
+
+    request.setMultiSignature({
+      signature,
+      identifier,
+    })
+
+    return request
+  }
+
+  private async signRequest<Request extends IndyVdrRequest>(request: Request, signingKey: Key) {
+    const signature = await this.agentContext.wallet.sign({
+      data: TypedArrayEncoder.fromString(request.signatureInput),
+      key: signingKey,
+    })
+
+    request.setSignature({
+      signature,
+    })
+
+    return request
+  }
+
+  public async endorseTransaction<Request extends IndyVdrRequest>(transaction: Request, endorserDid: string) {
+    const endorserSigningKey = await verificationKeyForIndyDid(this.agentContext, endorserDid)
+    const { namespaceIdentifier } = parseIndyDid(endorserDid)
+
+    const endorsedTransaction = await this.multiSignRequest(transaction, endorserSigningKey, namespaceIdentifier)
+    return endorsedTransaction
+  }
+
+  public async endorseCreateDidRequest(didAction: EndorseDidTxAction): Promise<EndorseDidTxAction> {
```

Comment:
It works we either have 

```ts
public async signRequest<Request extends IndyVdrRequest>(request: Request, submitterDid: string) {
    const signingKey = await verificationKeyForIndyDid(this.agentContext, submitterDid)

    const signature = await this.agentContext.wallet.sign({
      data: TypedArrayEncoder.fromString(request.signatureInput),
      key: signingKey,
    })

    request.setSignature({
      signature,
    })

    return request
  }

  public async endorseTransaction<Request extends IndyVdrRequest>(transaction: Request, endorserDid: string) {
    const endorserSigningKey = await verificationKeyForIndyDid(this.agentContext, endorserDid)
    const { namespaceIdentifier } = parseIndyDid(endorserDid)

    const endorsedTransaction = await this.multiSignRequest(transaction, endorserSigningKey, namespaceIdentifier)
    return endorsedTransaction
  }
```
  
 or 
 
 
 ``` ts
public async endorseTransaction2<Request extends IndyVdrRequest>(transaction: Request, endorserDid: string) {
    const endorserSigningKey = await verificationKeyForIndyDid(this.agentContext, endorserDid)
    const { namespaceIdentifier } = parseIndyDid(endorserDid)

    let endorsedTransaction: Request
    if (transaction.body.signature) {
      endorsedTransaction = await this.multiSignRequest(transaction, endorserSigningKey, namespaceIdentifier)
    } else {
      endorsedTransaction = await this.signRequest(transaction, endorserDid)
    }
    return endorsedTransaction
}
```

Reply:
Nice! I think there's something to say for both approaches. The second one hides the real working a bit more, so maybe signRequest is better?  However, personally I prefer to have a single method that just endorses the transaction. Whether the signature is applied different doesn't really matter from a caller's perspective.

What do you prefer?


For both approaches we probably need to make sure that the submitterDid  (i don't know if we can extract that) in case of the single signature is the same as the did we're going to sign with. So we don't have the case where the submitterDid is X while the did used to sign the request is Y